File: 1239-maximum-length-of-a-concatenated-string-with-unique-characters/1239-maximum-length-of-a-concatenated-string-with-unique-characters.cpp
class Solution {
public:
    int maxLength(vector<string>& arr) {
          return process(arr, "", 0);
    }
    int process(vector<string>& arr, string soFar, int index) {
        if (index > arr.size()) return 0;
        set<char> set;
        for(char c: soFar) {
            if(set.count(c)) return 0;
            set.insert(c);
        }
        int maxi = soFar.length();
        for(int i = index; i < arr.size(); i++) {
            maxi = max(maxi, process(arr, soFar + arr[i], i + 1));
        }
        return maxi;
    }
};

File: 1239-maximum-length-of-a-concatenated-string-with-unique-characters/1239-maximum-length-of-a-concatenated-string-with-unique-characters.java
class Solution {
    public int maxLength(List<String> arr) {
         return process(arr, "", 0);
    }
    int process(List<String> arr, String soFar, int index) {
        if (index > arr.size()) return 0;
        Set<Character> set = new HashSet<>();
        for(char c: soFar.toCharArray()) {
            if(set.contains(c)) return 0;
            set.add(c);
        }
        int max = soFar.length();
        for(int i = index; i < arr.size(); i++) {
            max = Math.max(max, process(arr, soFar + arr.get(i), i + 1));
        }
        return max;
    }
}